### PR TITLE
Add vibration

### LINKED ISSUE #3586: Add vibrate in mobile webbrowsers
We don't currently vibrate on mobile phones in a browser. This is only implemented for the iOS app.

https://developer.mozilla.org/en-US/docs/Web/API/Navigator/vibrate

### PR DESCRIPTION
**Pull request by bramkragten**
I don't have a device that supports `vibrate`, and can't find a list of patterns, maybe someone can make some nice patterns?

Closes https://github.com/home-assistant/home-assistant-polymer/issues/3586

### CODE REVIEW

File: src/util/haptics.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  success: [50, 50, 50],
+  warning: [100, 50, 100],
+  failure: [200, 100, 200],
+  light: [100],
+  medium: [200],
+  heavy: [300],
+  selection: [50],
+};
+
+export const haptic = (hapticType: HapticType) => {
```
Comment (balloob) [inline src/util/haptics.ts]: Let's not conflate concerns of requesting a vibration and vibrating.

Instead, we should just listen for haptic events on window and vibrate accordingly. Leave all places with forwardHaptic in there. That way things are a little more reusable.

Comment (bramkragten) [reply src/util/haptics.ts]: What is the right place to add the event listener?

File: src/util/haptics.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  warning: [100, 50, 100],
+  failure: [200, 100, 200],
+  light: [100],
+  medium: [200],
+  heavy: [300],
+  selection: [50],
+};
+
+export const handleHaptic = (hapticType: HapticType) => {
+  if (navigator.vibrate) {
```
Comment (balloob) [inline src/util/haptics.ts]: what about you add this check before you attach the listener? 😄

File: src/entrypoints/core.ts [hunk 3]
```diff
@@ -99,3 +100,5 @@ window.addEventListener("error", (e) => {
     });
   }
 });
+
+window.addEventListener("haptic", (e) => handleHaptic(e.detail));
```
Comment (balloob) [inline src/entrypoints/core.ts]: We should not add it to `core.ts`, that file should be kept as lean as possible with only bootstrapping the things that are essential for fetching the data. Add it to the Home Assistant state mixins.

File: src/util/haptics.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { HapticType } from "../data/haptics";
```
Comment (balloob) [inline src/util/haptics.ts]: Let's put the content of this file in `haptic-mixin.ts`

**Review by balloob — APPROVED** [review]
ok to merge when comment addressed.

**Pull request merged**